### ISSUE TITLE
travis-ci: linux clang compiler builds fail

### ISSUE THREAD

**Issue by tph-thuering** [SwissTPH/openmalaria#144]
> For example:
> https://travis-ci.org/SwissTPH/openmalaria/builds/80286565
Error: https://travis-ci.org/SwissTPH/openmalaria/jobs/80286567#L488

```
In file included from /usr/include/xsd/cxx/zc-istream.hxx:215:

/usr/include/xsd/cxx/zc-istream.txx:35:7: error: use of undeclared identifier
      'setg'
      setg (b, b, e);
      ^
/usr/include/xsd/cxx/zc-istream.txx:17:7: note: in instantiation of member
      function 'xsd::cxx::zc_streambuf<char>::init' requested here
      init ();
      ^
/usr/include/xsd/cxx/zc-istream.txx:64:11: note: in instantiation of member
      function 'xsd::cxx::zc_streambuf<char>::zc_streambuf' requested here
        : buf_ (str)
          ^
/usr/include/xsd/cxx/zc-istream.txx:81:11: note: in instantiation of member
      function 'xsd::cxx::zc_istream_base<char>::zc_istream_base' requested here
        : zc_istream_base<C> (str),
          ^
/usr/include/xsd/cxx/tree/parsing/int.hxx:69:23: note: in instantiation of
      member function 'xsd::cxx::zc_istream<char>::zc_istream' requested here
        zc_istream<C> is (tmp);
                      ^
/usr/include/xsd/cxx/tree/parsing/int.hxx:51:16: note: in instantiation of
      member function 'xsd::cxx::tree::traits<int, char, 0>::create' requested
      here
        return create (xml::transcode<C> (a.getValue ()), 0, f, c);
               ^
/home/travis/build/SwissTPH/openmalaria/build/schema/scenario.cpp:1811:42: note: 
      in instantiation of member function 'xsd::cxx::tree::traits<int, char,
      0>::create' requested here
        this->number_.set (NumberTraits::create (i, f, this));
                                         ^
/usr/lib/gcc/x86_64-linux-gnu/4.6/../../../../include/c++/4.6/streambuf:489:7: note: 
      must qualify identifier to find this declaration in dependent base class
      setg(char_type* __gbeg, char_type* __gnext, char_type* __gend)
      ^
1 error generated.
make[2]: *** [schema/CMakeFiles/schema.dir/scenario.cpp.o] Error 1
```

I suppose a header file needs to be included. Although OS X clang builds work and gcc on both systems too (although OS X links gcc symbolically by default to clang). 

I think it must have something to do with openmalaria's dependencies, scope or the xsd version we're using on travis-ci.

@nicolasreed do you like to investigate?

**Comment by nicolasreed** [comment]
Sure thing. I will let you know if I find anything interesting.

**Comment by nicolasreed** [comment]
It seems to be a problem with clang's "Two-Phase Name Lookup", described in detail in [this blog post](http://blog.llvm.org/2009/12/dreaded-two-phase-name-lookup.html).

The recommended fix in that blog post is to prepend `this` to calls such as `setg (b, b, e);` which would become `this->setg (b, b, e);`. Since these calls are within the xsd library, it isn't something we should have to do (although we could apply a patch on xsd if we wanted to). Nevertheless, I do not think applying a patch is of much benefit since these errors were fixed in the latest XSD release (4.0.0). It may be best (and easiest) to update the OpenMalaria xsd dependency to v4.0.0.

Thoughts?

**Comment by tph-thuering** [comment]
@nicolasreed: Great! I think the travis container uses xsd 3.2 from ubuntu.. I guess we can only wait until an upgrade to a more recent version is available. A workaround would probably be possible too, but this is not very important, since compiling works with clang on other systems.

**Comment by tph-thuering** [comment]
Try to install xsd/4.0.0 from ppa or vivid sources

**Comment by tph-thuering** [comment]
Source: https://launchpad.net/ubuntu/+source/xsd which would be the ppa?

**Comment by sgtbigman** [comment]
Unfortunately, in my testing of that repository, there is no package present in it.

However, I did do some further testing in a Ubuntu Precise VM. Using the official .deb on the [XSD download page](http://codesynthesis.com/products/xsd/download.xhtml), as well as using lib-xerces-c3 (since the latest XSD seems to require it), I was able to successfully build openmalaria using clang.

Unfortunately, when I went to test this with Travis CI, I forgot that we cannot utilize `sudo` since we are using the container-based infrastructure. This prevents installing the .deb package, although there may be another alternative I'm forgetting at the moment.

My only other thought was to utilize the 'debian-sid' or 'trusty' sources and see if a newer version of xsd is installable through one of them. I'll look more into this and keep you posted on my progress.

Thanks.

**Comment by tph-thuering** [comment]
This doc would come in handy: http://docs.travis-ci.com/user/migrating-from-legacy/#Adding-APT-Sources but I don't know yet which to add.

**Comment by tph-thuering** [comment]
@sgtbigman xsd 4.0 is in ubuntu from 14.10 [utopic](http://manpages.ubuntu.com/manpages/utopic/man1/xsdcxx.1.html). So maybe Debian unstable (sid) might be the way to go.

Dependencies on vivid (15.04) for package xsdcxx (4.0.0-1): 
> apt-cache show xsdcxx
> Depends: libc6 (>= 2.14), libgcc1 (>= 1:4.1.1), libstdc++6 (>= 4.9), libxerces-c3.1

Versions for sid might vary, but xsd upstream has the same version.

**Comment by sgtbigman** [comment]
In trying to utilize the xsdcxx package from the Debian sid repository, I started running into some libc errors. I may continue looking into this, but I figured it may be best to just wait for Travis CI to update the base image to a newer version of Ubuntu.

Thanks.

**Comment by dhardy** [comment]
Oh, this issue is a bug in xsdcxx. I reported it ages back... they didn't want to make a bugfix release.

**Comment by sgtbigman** [comment]
Right, I believe it's the same bug I mentioned previously: https://github.com/SwissTPH/openmalaria/issues/144#issuecomment-142592086. Since Travis CI is using older Ubuntu versions, the packaged xsdcxx version is not >= 4.x.x, which includes the fix for this bug.

Once Travis CI moves to newer versions of Ubuntu, this issue *should* resolve itself.

**Comment by sgtbigman** [comment]
Hello all,

I just ran a [successful test with clang on Linux](https://travis-ci.org/sgtbigman/openmalaria/jobs/113316427) of the latest commit using the ['Trusty beta build environment'](https://docs.travis-ci.com/user/trusty-ci-environment/).

Not sure it would make much sense to switch to using a beta-level feature just yet, but I wanted to let you know that once the Trusty build environment is out of beta, all 4 test builds should pass successfully.

Thanks, and congrats on the latest release! :+1:

**Comment by tph-thuering** [comment]
Sure, do a PR! The sooner the better. Trusty is already a few years old and is LTS, we're currently at wily http://releases.ubuntu.com/wily/ . We just do the test and then we merge it into production.
Make sure you catch up to the latest master commit for your fork. Best do a rebase on your branch so the git history stays linear when we merge into our master.

**Comment by tph-thuering** [comment]
@sgtbigman: Sure, do a PR! The sooner the better. Trusty is already a few years old and is LTS, we're currently at wily http://releases.ubuntu.com/wily/ . We just do the test and then we merge it into production.
Make sure you catch up to the latest master commit for your fork. Best do a rebase on your branch so the git history stays linear when we merge into our master.

**Comment by tph-thuering** [comment]
@sgtbigman: You need to thank @dhardy for the release as he implemented all the new features. I just did some cleaning.

**Comment by dhardy** [comment]
No problem. I don't know why you need a beta environment; I've been using Clang (off and on) for a couple of years now.

**Comment by tph-thuering** [comment]
@dhardy I think he refers to the travis "beta", which uses trusty images instead of the older ubuntu distribution which provides older libraries, which is the source of the compiler build failures with clang, see here for example: https://travis-ci.org/SwissTPH/openmalaria/jobs/112681265#L545

**Comment by tph-thuering** [comment]
@dhardy It uses xsd 3.3, the build error is fixed with xsd 4.0 which is available with ubuntu trusty (aka beta environment for travis)

**Comment by tph-thuering** [comment]
@dhardy It uses xsd 3.3, the build error is fixed with xsd 4.0 which is available with ubuntu trusty (aka beta environment for travis)

see here: https://travis-ci.org/SwissTPH/openmalaria/jobs/112681265#L482

**Comment by tph-thuering** [comment]
@sgtbigman you can also use the dedicated branch for this bug: [fix-clang-linux-builds#144](https://github.com/SwissTPH/openmalaria/tree/fix-clang-linux-builds%23144) for fixing this bug. But I can do it if you prefer as it is a small change.

**Comment by dhardy** [comment]
@tph-thuering XSD 3.3 works with a one-line fix. IIRC it's as simple as adding `this->` before some variable. But it's annoying to have to patch system packages, sure. I don't have anything against using a beta Ubuntu release for compiling.

Actually, I kind of wonder if providing Linux binaries is even important, given that the main Linux machines running it will be clusters which will probably have custom-built binaries anyway.

**Comment by tph-thuering** [comment]
nice, it worked: https://travis-ci.org/SwissTPH/openmalaria/jobs/113419318

**Comment by tph-thuering** [comment]
Will merge this evening to develop.

**Comment by sgtbigman** [comment]
I believe this has been resolved in #157. Feel free to let me know otherwise :)
